### PR TITLE
Introduce RocksDbSegmentIdentifier to avoid changing the storege plug…

### LINKED ISSUE #3722: Segfault on latest besu, org.rocksdb.WriteOptions
### Description
There have been 3+ recent reports of segfaults on besu, all of them in the rocksdb codebase.  These reports have come in since the upgrade to rocksdbjni 6.29.5

see the segfault log outputs:
https://gist.github.com/garyschulte/c45d068399c6309e0173a0dae09979b8

### Acceptance Criteria
* reproduce segfault reliably
* address the segfault

### Steps to Reproduce (Bug)
1. ?
**Expected behavior:** [What you expect to happen]

**Actual behavior:** [What actually happens]

**Frequency:** [What percentage of the time does it occur?]

### Versions (Add all that apply)
* Software version: besu-22.4.0-RC2-SNAPSHOT +
* Java version: reported on java 11, java 17
* OS Name & Version: windows and linux amd64

### Additional Information (Add any of the following or anything else that may be relevant)
* Besu setup info - genesis file, config options
* System info - memory, CPU
[Uploading... hs_err_pid8.log]

### PR DESCRIPTION
**Pull request by fab-10**
…in interface

Signed-off-by: Fabio Di Fabio <fabio.difabio@consensys.net>

## PR description

A refactor of #3734 to avoid changing the storage plugin interface

## Fixed Issue(s)

fixes #3722 

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.rocksdb.ColumnFamilyDescriptor;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.TransactionDB;
+
+public class RocksDbSegmentIdentifier {
```
Comment (garyschulte) [inline plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java:27]: how about RocksDbColumnFamilyIdentifier instead?

File: plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  private final TransactionDB db;
+  private final AtomicReference<ColumnFamilyHandle> reference;
+
+  public RocksDbSegmentIdentifier(
+      final TransactionDB db, final ColumnFamilyHandle columnFamilyHandle) {
+    this.db = db;
+    this.reference = new AtomicReference<>(columnFamilyHandle);
+  }
+
+  public void recreate() {
```
Comment (garyschulte) [inline plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java:38]: I don't think recreate implies storage is reclaimed.  how about reset or clear?

**Review by garyschulte — APPROVED** [review]
Love it.  Feedback on naming, but nothing blocking

File: plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java (hunk 2 above)
Comment (fab-10) [reply plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java:38]: `reset` sound good

File: plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java (hunk 1 above)
Comment (fab-10) [reply plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/RocksDbSegmentIdentifier.java:27]: Used SegmentedIdentifier to follow the naming used in the interface, but I do not have a strong opinion on this, so if you think `RocksDbColumnFamilyIdentifier` is fits better, I can rename it

File: plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java [hunk 3]
```diff
@@ -150,42 +149,41 @@ private BlockBasedTableConfig createBlockBasedTableConfig(final RocksDBConfigura
   }
 
   @Override
-  public AtomicReference<ColumnFamilyHandle> getSegmentIdentifierByName(
-      final SegmentIdentifier segment) {
+  public RocksDbSegmentIdentifier getSegmentIdentifierByName(final SegmentIdentifier segment) {
```
Comment (matkt) [inline plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java:152]: is this modification voluntary to revert what did @garyschulte on this PR ? https://github.com/hyperledger/besu/pull/3734/files#diff-1b1c62528a9b69821bd095065e34f149b9be0f6abdc301d86e7b9dbf14fc3318R154

Comment (fab-10) [reply plugins/rocksdb/src/main/java/org/hyperledger/besu/plugin/services/storage/rocksdb/segmented/RocksDBColumnarKeyValueStorage.java:152]: yes, this revert the use of AtomicReference<S> in the interface, and keeps the storage plugin interface stable

**Review by matkt — APPROVED** [review]
LGTM

**Pull request merged**